### DISCUSSION TITLE
Creating a readable stream from async iterables

### DISCUSSION THREAD

**Discussion by ducaale** [denoland/deno#8808] [Q&A] [answered]
Is there an equivalent to Node's `Readable.from()` in Deno?

```javascript
import {Readable} from 'stream';

function* ids() {
  for (const i = 0; i < 10; i++) {
    yield i;
  }
}

Readable.from(ids());
```

**Comment by kitsonk** [comment]
Deno has WHATWG streams built in: https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream. But this might be overkill, and so there is also the `BufReader`: https://doc.deno.land/https/deno.land/std/io/bufio.ts#BufReader

**Comment by ducaale** [comment]
Relevant [issue](https://github.com/whatwg/streams/issues/1018) and [draft](https://github.com/whatwg/streams/pull/1083) for supporting `ReadableStream.from()` in WHATWG streams.

My use-case was to create a readable stream from an async-iterable so it can be returned as the body of an HTTP response using https://github.com/oakserver/oak. Is there an alternative/more idiomatic approach in Deno?

**Comment by danopia** [comment, marked as answer]
Perhaps this std import/function `readableStreamFromAsyncIterator` answers your original question: https://deno.land/std@0.81.0/io/streams.ts#L50

```typescript
async function* ids() {
  for (let i = 0; i < 10; i++) {
    yield i;
  }
}

// Create a stream
import {readableStreamFromAsyncIterator} from "https://deno.land/std@0.81.0/io/streams.ts";
const idStream = readableStreamFromAsyncIterator(ids());

// Consume the stream
import {collect} from "https://uber.danopia.net/deno/observables-with-streams@v1/sinks/collect.ts";
console.log('All IDs:', await collect(idStream));
```

As long as you're talking about **async** iterables, that is. If your iterable isn't async then std doesn't have an implementation. The non-async implementation is very basic though: [from-iterable.ts](https://github.com/danopia/observables-with-streams/blob/master/src/sources/from-iterable.ts)
